### ISSUE TITLE
Extend Teleport to OAuth2 apps (Grafana, Happa, Prometheus, etc.)

### ISSUE THREAD

**Issue by gawertm** [giantswarm/roadmap#3035]

**Comment by tuladhar** [comment]
Discusssion topics:
- Teleport application access conflicts with Dex.
- Grafana is used by customer, which implies customer will have to use Teleport to access the Grafana.
- Happa - Teleport does not handle user session, similar to how Dex does, for e.g: who logged in as.

**Comment by tuladhar** [comment]
Prometheus access via Teleport is working:
- https://prometheus-golem.teleport.giantswarm.io/golem/graph?g0.expr=&g0.tab=1&g0.stacked=0&g0.show_exemplars=0&g0.range_input=1h

**Comment by tuladhar** [comment]
Document explaining Prometheus access setup:

- https://docs.google.com/document/d/1t3YnhUmu0UOSNhw5kr8GBLRJjuZcDf_kGXXlaAP7dDE/edit#heading=h.8vev9pjz9n7p

**Comment by tuladhar** [comment]
![Image](https://github.com/giantswarm/roadmap/assets/5674762/2f37dbb9-dea5-47cc-b16c-3ecd86500f1d)

**Comment by tuladhar** [comment]
Plan B worked - [grafana-golem.teleport.giantswarm.io](http://grafana-golem.teleport.giantswarm.io/)

Plan A didn’t work, and is not backward compatible with customer, as there are breaking changes when setting up dex behind Teleport.

**Why Plan A didn't work?**
- Starting off, we have to update `GF_AUTH_GENERIC_OAUTH_AUTH_URL=http://dex-golem.teleport.giantswarm.io/` for Grafana, otherwise, it will keep redirecting us to VPN accessible endpoint [dex.golem.gaws.gigantic.io](http://dex.golem.gaws.gigantic.io/)
-   After that, dex login works using Teleport-powered Dex endpoint, but in Grafana, we get `Login failed: Missing saved oauth state`.

![Image](https://github.com/giantswarm/roadmap/assets/5674762/a2bd9935-1c7e-4828-8b3f-9e50ae5008d9)

- And to fix that, we need to update Grafana `GF_SERVER_ROOT_URL` env to `https://grafana-golem.teleport.giantswarm.io/`

- But, after changing `GF_SERVER_ROOT_URL` we hit this issue:

![Image](https://github.com/giantswarm/roadmap/assets/5674762/24d31ee2-face-485f-9c34-ed29170fd7c3)

- To fix that, we need to update dex-operator secret to add `https://grafana-golem.teleport.giantswarm.io/` to list of redirectURIs:
```
 redirectURIs:
 - https://grafana.golem.gaws.gigantic.io/login/generic_oauth
 - https://grafana-golem.teleport.giantswarm.io/login/generic_oauth
 name: grafana
```

- Then, we get Grafana with Dex login behind Teleport. But, with all those changes, dex login for customer breaks.

**How Plan B worked?**

- With B, no changes needed on dex side, and normal customer dex login flow works the same. The only change happens at Grafana side by adding this to configmap:
```
    [auth.jwt]
    enabled = true
    auto_sign_up = true
    header_name = Teleport-Jwt-Assertion
    username_claim = sub
    jwk_set_url = https://teleport.giantswarm.io/.well-known/jwks.json
```

- All request from Teleport, includes JWT token pass-through and sent directly to Grafana pod. 
- Grafana access via existing ingress  [grafana.golem.gaws.gigantic.io](http://grafana.golem.gaws.gigantic.io/) works the same.

**Comment by anvddriesch** [comment]
Since customers dex access can not be affected by either scenario, we are unable to change settings like the OIDC issuer flags on the kubernetes API server or the oauth provider flags for grafana. (both allow only one issuer)
This is not an issue in scenario B since we can use the jwt token directly and bypass dex completely.
However, it is an issue in scenario A for the reasons described in Purus comment. 
The limitation on issuers is the reason we were using dex in the first place. We may have a bit more freedom with 1.29 structured auth config options but we have not tested this yet.
As of right now, running dex inside teleport to make it available to us does not seem possible if we don't also want to make dex unavailable to customers.
We already have a limitation on reproducing dex issues due to identity provider configuration on the customers side. Bypassing dex and removing the VPN will grow this limitation since dex will become unavailable to us. However, since we already need to debug in tandem with customers it may not be a huge change.

**Comment by tuladhar** [comment]
### Feedback from Dmitry:
- **How does local setup works with Teleport?**
  - **Resolved:** By running local teleport app agent pointing to local Happa instance, confirmed with Dmitry, it worked.

-   **Is there a way to get a token not from a header? Separate api endpoint maybe? the problem with the header is that we cannot access it from client side (from JS). We would need to have a server to transfer it to the client, but we don't have a server for happa.**
     - Teleport automatically injects JWT tokens into the headers of requests going through the Application Access proxy, **it doesn't seem to provide an endpoint for directly accessing the JWT token**. There is only JWKS endpoint for verifying the token - https://teleport.giantswarm.io/.well-known/jwks.json. 
  - [Asked in Teleport community Slack channel.](https://goteleport.slack.com/archives/CEZH6UL64/p1706776226038469)

**Comment by tuladhar** [comment]
### Atlas Feedback (Grafana/Loki)
- If I use grafana through teleport, the loki datasource which forward the OAuth token is failing with the current loki-multi-tenant-proxy implementation. On the other hand, if I use grafana through https://grafana.golem.gaws.gigantic.io/ it works fine. Which OAuth token are you provided to grafana ? azure AD? [(Thread)](https://gigantic.slack.com/archives/C01176DKNP4/p1706698088051739?thread_ts=1706693167.306219&cid=C01176DKNP4)
  - Grafana access via Teleport, doesn't send oAuth token instead JWT token is sent, as we don't go through Dex.
  - I have setup a call with Atlas to discuss more on this.

**Comment by tuladhar** [comment]
**Challenge: Loki DS uses oAuth token forward**
- **Problem:** Since, login to Grafana uses Teleport JWT token, oAuth token setup for Loki DS breaks.
- **Potential solution:** Forward `Teleport-Jwt-Assertion` header from Loki DS. Atlas is looking into it.

**Comment by tuladhar** [comment]
Honeybadger - `Challenge #1` - Happa can't accept Teleport Jwt token
- **Problem** Since, Happa is fully client side without server component, Dmitry confirmed that it can't access Teleport JWT token from the header passed on by Teleport. And, teleport doesn't offer an endpoint to access the JWT token directly.
- **Potential solution:** Re-configure nginx with an endpoint `/jwt-token` that reads the `Teleport-Jwt-Assertion` header and return to response.

**Comment by tuladhar** [comment]
## Discussion with Honeybadger

Slides: https://docs.google.com/presentation/d/1OLtCijqp6VM1Xu2p0sxvyr1d2NJD6wHK0ypQ1mzRRCA/edit?usp=sharing

**Action Items:**
- BigMac to delivery https://github.com/giantswarm/roadmap/issues/3220 and Honeybadger can then start experimenting and make changes to Happa.
  - Dmitry, will look into webpack webserver to expose the JWT token endpoint for local development. We will assist if help is needed.
  - BigMac to setup happaapi behind Teleport, and Honeybadger will start experiementing and make changes to Happa API.
- DevPortal is low priority, and for now, we won't move it behind Teleport until we're done with Happa.
- BigMac to update opsctl to adjust `opsctl open`, where for CAPI it should go through teleport powered happa endpoint (e.g: happa-golem.teleport.giantswarm.io), and for others, dex powered happa.

**Comment by tuladhar** [comment]
## Discussion with Dmitry
**Attendees:** Puru, Spyros, Antonia, Pawel, Dmitry

**Action Items:**
- BigMac blocked on Happa teleport integration due to Happa K8s API access
- BigMac blocked on Backstage teleport integration due to Dex
- Dmitry to discuss with honeybadger about “Happa backend” and research on “OIDC” vs “service account” for Backstage K8s API access, Backstage permission framework.
- BigMac to help with the review.

**Comment by tuladhar** [comment]
## SOCKS5 proxy as an alternative

It seems we can leverage the SOCKS5 proxy feature of Teleport SSH to access internal web apps. [I have documented the setup process in intranet](https://intranet.giantswarm.io/docs/support-and-ops/teleport/web-access/). I posted the [demo in our channel](https://gigantic.slack.com/archives/C053JHJC99Q/p1708946864174769).

**Comment by tuladhar** [comment]
@gawertm Here's the network diagram. Let me know if this doesn't make sense :) 
![Screenshot 2024-02-29 at 9 56 41 PM](https://github.com/giantswarm/roadmap/assets/5674762/87e3b1a2-c40f-4bef-ab8c-b4cee94cec99)

> Only use proxy for accessing internal web apps, because potentially, we can use it to browse other external websites.

Yes, think of SOCKS5 proxy as VPN through SSH connection. So, any network access the node where SOCKS5 is established on, we can access them too, which means access to the internet if node has access to it along with internal network access.

**Comment by gawertm** [comment]
ok understood, thanks! lets discuss it in the refinement today :)

**Comment by gawertm** [comment]
we came to the conclusion, that this solution is not user friendly enough and too much config is needed upfront. While it is a good emergency access solution, we will still keep the VPN for App access in private installation.
to be confirmed with @alex-dabija. Also we will revisit with kubernetes 1.29 and structured auth config

**Comment by alex-dabija** [comment]
We agreed in SIG Architecture to keep the the VPN for now.

**Comment by stone-z** [comment]
The latest (v16) release of teleport now support the VNET feature, which has enabled app access. Grafana, happa, and the Kyverno UI are currently supported.
